File: lib/core/domain/models/controversy.dart
/// Domain model for a single documented controversy / criticism episode
/// involving a public figure.
///
/// Controversies are produced by the language model as part of the
/// biography call (`functions/lib/groq.js`), filtered by the
/// corroboration gate so that a severe claim no retrieved article
/// supports is discarded rather than rendered, and shown in the
/// Controversy section of the dashboard. Unlike the earlier flat
/// `List<String>`, each entry is structured so the UI can sort by
/// severity, group by category, and show status at a glance.
library;

import 'package:equatable/equatable.dart';

/// Broad category buckets a controversy can fall into. Kept as a small
/// closed set so the UI can assign consistent colours and icons.
abstract final class ControversyCategory {
  static const legal = 'Legal';
  static const financial = 'Financial';
  static const socialMedia = 'Social media';
  static const personalConduct = 'Personal conduct';
  static const political = 'Political';
  static const professional = 'Professional';
  static const relationships = 'Relationships';
  static const other = 'Other';

  static const all = [
    legal,
    financial,
    socialMedia,
    personalConduct,
    political,
    professional,
    relationships,
    other,
  ];

  /// Normalises an arbitrary model-supplied string to one of [all],
  /// falling back to [other] for anything unrecognised.
  static String normalize(String? raw) {
    if (raw == null) return other;
    final lower = raw.toLowerCase();
    for (final c in all) {
      if (c.toLowerCase() == lower) return c;
    }
    // Loose keyword matching for near-misses.
    if (lower.contains('law') ||
        lower.contains('court') ||
        lower.contains('lawsuit')) {
      return legal;
    }
    if (lower.contains('money') ||
        lower.contains('tax') ||
        lower.contains('fraud')) {
      return financial;
    }
    if (lower.contains('tweet') ||
        lower.contains('post') ||
        lower.contains('online')) {
      return socialMedia;
    }
    if (lower.contains('politic') || lower.contains('election')) {
      return political;
    }
    if (lower.contains('work') ||
        lower.contains('career') ||
        lower.contains('set')) {
      return professional;
    }
    if (lower.contains('relationship') ||
        lower.contains('divorce') ||
        lower.contains('affair')) {
      return relationships;
    }
    return other;
  }
}

/// Lifecycle status of a controversy.
abstract final class ControversyStatus {
  static const ongoing = 'ongoing';
  static const resolved = 'resolved';
  static const historical = 'historical';

  static String normalize(String? raw) {
    final lower = (raw ?? '').toLowerCase();
    if (lower.contains('ongoing') ||
        lower.contains('active') ||
        lower.contains('current')) {
      return ongoing;
    }
    if (lower.contains('resolved') ||
        lower.contains('settled') ||
        lower.contains('over')) {
      return resolved;
    }
    return historical;
  }
}

class Controversy extends Equatable {
  const Controversy({
    required this.title,
    required this.summary,
    required this.category,
    required this.severity,
    required this.status,
    this.year,
    this.sources = const [],
  });

  /// Short headline for the episode (≤ ~10 words).
  final String title;

  /// 1–3 sentence neutral description of what happened.
  final String summary;

  /// One of [ControversyCategory.all].
  final String category;

  /// 1 (minor backlash) … 5 (major scandal with lasting consequences).
  final int severity;

  /// One of [ControversyStatus] values.
  final String status;

  /// Approximate year the episode began, when known.
  final int? year;

  /// Publication names or URLs backing the entry.
  final List<String> sources;

  bool get isOngoing => status == ControversyStatus.ongoing;

  @override
  List<Object?> get props => [title, category, severity, status, year];

  Map<String, dynamic> toMap() => {
    'title': title,
    'summary': summary,
    'category': category,
    'severity': severity,
    'status': status,
    if (year != null) 'year': year,
    'sources': sources,
  };

  factory Controversy.fromMap(Map<String, dynamic> map) {
    return Controversy(
      title:
          (map['title'] as String?)?.trim().isNotEmpty == true
              ? (map['title'] as String).trim()
              : 'Untitled controversy',
      summary: (map['summary'] as String?)?.trim() ?? '',
      category: ControversyCategory.normalize(map['category'] as String?),
      severity: ((map['severity'] as num?)?.toInt() ?? 1).clamp(1, 5),
      status: ControversyStatus.normalize(map['status'] as String?),
      year: (map['year'] as num?)?.toInt(),
      sources:
          (map['sources'] as List<dynamic>?)
              ?.map((e) => e.toString().trim())
              .where((s) => s.isNotEmpty)
              .toList() ??
          const [],
    );
  }
}

File: site/lib/controversy.ts
/**
 * Structured controversy record.
 *
 * This is the web twin of `lib/core/domain/models/controversy.dart`. It
 * exists so the marketing/reference site can render the same typed,
 * sourced, dated, severity-scored records the app does, and feed them to
 * the shared Controversy Index in `controversy-index.ts`.
 *
 * Keep the category list, the status list, and both `normalize` functions
 * in step with the Dart file. A change here needs the same change there,
 * plus a matching test case in both suites.
 */

export const CONTROVERSY_CATEGORIES = [
  "Legal",
  "Financial",
  "Social media",
  "Personal conduct",
  "Political",
  "Professional",
  "Relationships",
  "Other",
] as const;

export type ControversyCategory = (typeof CONTROVERSY_CATEGORIES)[number];

export const CONTROVERSY_STATUSES = [
  "ongoing",
  "resolved",
  "historical",
] as const;

export type ControversyStatus = (typeof CONTROVERSY_STATUSES)[number];

export interface Controversy {
  /** Short headline for the episode. */
  title: string;
  /** One to three neutral sentences describing what happened. */
  summary: string;
  category: ControversyCategory;
  /** 1 (minor backlash) to 5 (major scandal with lasting consequences). */
  severity: number;
  status: ControversyStatus;
  /** Approximate year the episode began, when known. */
  year?: number;
  /** Publication names or URLs backing the entry. */
  sources: string[];
}

/**
 * Normalises an arbitrary category string to one of
 * {@link CONTROVERSY_CATEGORIES}, falling back to "Other". Mirrors
 * `ControversyCategory.normalize` in the Dart model, including the loose
 * keyword matching for near-misses.
 */
export function normalizeCategory(raw?: string | null): ControversyCategory {
  if (raw == null) return "Other";
  const lower = raw.toLowerCase();
  for (const c of CONTROVERSY_CATEGORIES) {
    if (c.toLowerCase() === lower) return c;
  }
  if (
    lower.includes("law") ||
    lower.includes("court") ||
    lower.includes("lawsuit")
  ) {
    return "Legal";
  }
  if (
    lower.includes("money") ||
    lower.includes("tax") ||
    lower.includes("fraud")
  ) {
    return "Financial";
  }
  if (
    lower.includes("tweet") ||
    lower.includes("post") ||
    lower.includes("online")
  ) {
    return "Social media";
  }
  if (lower.includes("politic") || lower.includes("election")) {
    return "Political";
  }
  if (
    lower.includes("work") ||
    lower.includes("career") ||
    lower.includes("set")
  ) {
    return "Professional";
  }
  if (
    lower.includes("relationship") ||
    lower.includes("divorce") ||
    lower.includes("affair")
  ) {
    return "Relationships";
  }
  return "Other";
}

/** Mirrors `ControversyStatus.normalize` in the Dart model. */
export function normalizeStatus(raw?: string | null): ControversyStatus {
  const lower = (raw ?? "").toLowerCase();
  if (
    lower.includes("ongoing") ||
    lower.includes("active") ||
    lower.includes("current")
  ) {
    return "ongoing";
  }
  if (
    lower.includes("resolved") ||
    lower.includes("settled") ||
    lower.includes("over")
  ) {
    return "resolved";
  }
  return "historical";
}

export function isOngoing(c: Controversy): boolean {
  return c.status === "ongoing";
}

/**
 * Parses an untrusted record into a {@link Controversy}, clamping severity
 * to 1..5 and normalising the enums. Mirrors `Controversy.fromMap`.
 */
export function parseControversy(raw: Record<string, unknown>): Controversy {
  const title =
    typeof raw.title === "string" && raw.title.trim().length > 0
      ? raw.title.trim()
      : "Untitled controversy";
  const summary =
    typeof raw.summary === "string" ? raw.summary.trim() : "";
  const severityRaw =
    typeof raw.severity === "number" ? Math.trunc(raw.severity) : 1;
  const severity = Math.min(5, Math.max(1, severityRaw));
  const year =
    typeof raw.year === "number" ? Math.trunc(raw.year) : undefined;
  const sources = Array.isArray(raw.sources)
    ? raw.sources.map((s) => String(s).trim()).filter((s) => s.length > 0)
    : [];
  return {
    title,
    summary,
    category: normalizeCategory(
      typeof raw.category === "string" ? raw.category : null,
    ),
    severity,
    status: normalizeStatus(
      typeof raw.status === "string" ? raw.status : null,
    ),
    year,
    sources,
  };
}

/**
 * The corroboration gate, mirrored for the site. A severity 4 or 5 claim
 * with no backing source is dropped rather than rendered, exactly as the
 * backend's `corroborate.js` does before the record is ever stored.
 */
export function passesCorroborationGate(c: Controversy): boolean {
  if (c.severity >= 4) return c.sources.length > 0;
  return true;
}

/** Applies {@link passesCorroborationGate} across a list. */
export function corroborated(items: Controversy[]): Controversy[] {
  return items.filter(passesCorroborationGate);
}
